runtime/pack/dist/opt/termdebug/plugin/termdebug.vim

```diff
@@ -228,85 +540,352 @@ func s:Evaluate(range, arg)
', '\1', '')
+  if s:evalFromBalloonExpr
+    if s:evalFromBalloonExprResult == ''
+      let s:evalFromBalloonExprResult = s:evalexpr . ': ' . value
+    else
+      let s:evalFromBalloonExprResult .= ' = ' . value
+    endif
+    call s:OpenHoverPreview([s:evalFromBalloonExprResult], v:null)
+  else
+    echomsg '"' . s:evalexpr . '": ' . value
+  endif
+
+  if s:evalexpr[0] != '*' && value =~ '^0x' && value != '0x0' && value !~ '"$'
+    " Looks like a pointer, also display what it points to.
+    let s:ignoreEvalError = 1
+    call s:SendEval('*' . s:evalexpr)
+  else
+    let s:evalFromBalloonExprResult = ''
+  endif
 endfunc
 
+function! s:ShouldUseFloatWindow() abort
+  if has('nvim_open_win') && exists('g:termdebug_useFloatingHover') && (g:termdebug_useFloatingHover == 1)
+    return v:true
+  else
+    return v:false
+  endif
+endfunction
+
+function! s:CloseFloatingHoverOnCursorMove(win_id, opened) abort
+  if getpos('.') == a:opened
+    " Just after opening floating window, CursorMoved event is run.
+    " To avoid closing floating window immediately, check the cursor
+    " was really moved
+    return
+  endif
+  autocmd! plugin-LC-neovim-close-hover
+  let winnr = win_id2win(a:win_id)
+  if winnr == 0
+    return
+  endif
+  execute winnr . 'wincmd c'
+endfunction
+
+function! s:CloseFloatingHoverOnBufEnter(win_id, bufnr) abort
+    let winnr = win_id2win(a:win_id)
+    if winnr == 0
+        " Float window was already closed
+        autocmd! plugin-LC-neovim-close-hover
+        return
+    endif
+    if winnr == winnr()
+        " Cursor is moving into floating window. Do not close it
+        return
+    endif
+    if bufnr('%') == a:bufnr
+        " When current buffer opened hover window, it's not another buffer. Skipped
+        return
+    endif
+    autocmd! plugin-LC-neovim-close-hover
+    execute winnr . 'wincmd c'
+  endfunction
+
+" Open preview window. Window is open in:
+"   - Floating window on Neovim (0.4.0 or later)
+"   - Preview window on Neovim (0.3.0 or earlier) or Vim
+function! s:OpenHoverPreview(lines, filetype) abort
+    " Use local variable since parameter is not modifiable
+    let lines = a:lines
+    let bufnr = bufnr('%')
+
+    let use_float_win = s:ShouldUseFloatWindow()
+    if use_float_win
+        let bufname = nvim_create_buf(v:false, v:true)
+        call nvim_buf_set_lines(buf, 0, -1, v:true, lines)
+        let pos = getpos('.')
+
+        " Calculate width and height and give margin to lines
+        let width = 0
+        for index in range(len(lines))
+            let line = lines[index]
+            if line !=# ''
+                " Give a left margin
+                let line = ' ' . line
+            endif
+            let lw = strdisplaywidth(line)
+            if lw > width
+                let width = lw
+            endif
+            let lines[index] = line
+        endfor
+
+        " Give margin
+        let width += 1
+        let lines = [''] + lines + ['']
```

Comment:
Vertical margin of an entire screen row consumes quite a lot of screen space. Maybe we can remove this for now, and soon we can make this configurable by user with the 'margin' option.